sdk/trace/src/test/scala/org/typelevel/otel4s/sdk/trace/SdkTraceScopeSuite.scala

```diff
@@ -0,0 +1,310 @@
+        _ <- assertIO(scope.current, None)
+      } yield ()
+    }
+  }
+
+  test("childScope - use invalid context when there is an valid SpanContext") {
+    val invalid = SpanContext.invalid
+
+    PropF.forAllF { (valid: SpanContext) =>
+      for {
+        scope <- createTraceScope
+
+        _ <- assertIO(scope.current, None)
+
+        scope1 <- scope.childScope(invalid)
+        _ <- scope1 {
+          for {
+            _ <- assertIO(scope.current, Some(invalid))
+            scope2 <- scope.childScope(valid)
+            _ <- scope2(assertIO(scope.current, Some(invalid)))
+            _ <- assertIO(scope.current, Some(invalid))
+          } yield ()
+        }
+
+        // the context must be reset
+        _ <- assertIO(scope.current, None)
+      } yield ()
+    }
+  }
+
+  test("noopScope - always use invalid context") {
+    val invalid = SpanContext.invalid
+
+    PropF.forAllF { (ctx: SpanContext) =>
+      for {
+        scope <- createTraceScope
+
+        _ <- assertIO(scope.current, None)
+
+        // the context is empty
+        _ <- assertIO(scope.noopScope(scope.current), Some(invalid))
+
+        scope1 <- scope.childScope(ctx)
+
+        // the context has a valid span
+        _ <- scope1 {
+          for {
+            _ <- assertIO(scope.current, Some(ctx))
+            _ <- assertIO(scope.noopScope(scope.current), Some(invalid))
+            _ <- assertIO(scope.current, Some(ctx))
+          } yield ()
+        }
+
+        // the context must be reset
+        _ <- assertIO(scope.current, None)
+      } yield ()
+    }
+  }
+
+  test("withContext - use the given context") {
+    PropF.forAllF { (value: String) =>
+      for {
+        scope <- createTraceScope
+        key <- Context.Key.unique[IO, String]("some-key")
+
+        _ <- assertIO(scope.contextReader(identity), Context.root)
+
+        _ <- scope.withContext(Context.root.updated(key, value)) {
+          assertIO(
+            scope.contextReader(_.updated(key, value)).map(_.get(key)),
+            Some(value)
+          )
+        }
+
+        // the context must remain unchanged
+        _ <- assertIO(scope.contextReader(identity), Context.root)
+      } yield ()
+    }
+  }
+
+  test("contextReader - use the given context") {
+    PropF.forAllF { (value: String) =>
+      for {
+        scope <- createTraceScope
+        key <- Context.Key.unique[IO, String]("some-key")
+
+        _ <- assertIO(scope.contextReader(identity), Context.root)
+
+        _ <- assertIO(
+          scope.contextReader(_.updated(key, value)).map(_.get(key)),
+          Some(value)
+        )
+
+        // the context must remain unchanged
+        _ <- assertIO(scope.contextReader(identity), Context.root)
+      } yield ()
+    }
+  }
+
+  private def createTraceScope: IO[SdkTraceScope[IO]] =
+    IOLocal(Context.root).map { implicit ioLocal =>
+      SdkTraceScope.fromLocal[IO]
+    }
+
+  private implicit def localForIoLocal[F[_]: MonadCancelThrow: LiftIO, E](
```

Comment:
currently we have this method in the java-common module as well. should we promote it to core?

Reply:
Sounds reasonable